BParkingNano/plugins/KstarBuilder.cc

```diff
@@ -0,0 +1,174 @@
+#include "helper.h"
+#include <limits>
+#include <algorithm>
+#include "KinVtxFitter.h"
+
+
+class KstarBuilder : public edm::global::EDProducer<> {
+
+  // perhaps we need better structure here (begin run etc)
+public:
+  typedef std::vector<reco::TransientTrack> TransientTrackCollection;
+  
+
+  explicit KstarBuilder(const edm::ParameterSet &cfg):
+    trk1_selection_{cfg.getParameter<std::string>("trk1Selection")},
+    trk2_selection_{cfg.getParameter<std::string>("trk2Selection")},
+    pre_vtx_selection_{cfg.getParameter<std::string>("preVtxSelection")},
+    post_vtx_selection_{cfg.getParameter<std::string>("postVtxSelection")},
+    pfcands_{consumes<pat::CompositeCandidateCollection>( cfg.getParameter<edm::InputTag>("pfcands") )},
+    ttracks_{consumes<TransientTrackCollection>( cfg.getParameter<edm::InputTag>("transientTracks") )} {
+       produces<pat::CompositeCandidateCollection>();
+    }
+
+  ~KstarBuilder() override {}
+  
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions) {}
+  
+private:
+  const StringCutObjectSelector<pat::CompositeCandidate> trk1_selection_; // cuts on leading cand
+  const StringCutObjectSelector<pat::CompositeCandidate> trk2_selection_; // sub-leading cand
+  const StringCutObjectSelector<pat::CompositeCandidate> pre_vtx_selection_; // cut on the di-lepton before the SV fit
+  const StringCutObjectSelector<pat::CompositeCandidate> post_vtx_selection_; // cut on the di-lepton after the SV fit
+  const edm::EDGetTokenT<pat::CompositeCandidateCollection> pfcands_;
+  const edm::EDGetTokenT<TransientTrackCollection> ttracks_;
+};
+
+void KstarBuilder::produce(edm::StreamID, edm::Event &evt, edm::EventSetup const &) const {
+
+  //input
+  
+  edm::Handle<pat::CompositeCandidateCollection> pfcands;
+  evt.getByToken(pfcands_, pfcands);
+  
+  edm::Handle<TransientTrackCollection> ttracks;
+  evt.getByToken(ttracks_, ttracks);
+ 
+
+  // output
+  std::unique_ptr<pat::CompositeCandidateCollection> kstar_out(new pat::CompositeCandidateCollection());
+
+
+  //two mass hypothesis for tracks; we cannot discriminate which trak is K and which is pi
+  std::vector<std::pair<float, float>> mhypothesis{ std::make_pair(K_MASS,PI_MASS), std::make_pair(PI_MASS,K_MASS)  };
+  
+  for(size_t trk1_idx = 0; trk1_idx < pfcands->size(); ++trk1_idx) {
+    edm::Ptr<pat::CompositeCandidate> trk1_ptr(pfcands, trk1_idx);
+    if(!trk1_selection_(*trk1_ptr)) continue; 
+     
+    for(size_t trk2_idx = trk1_idx + 1; trk2_idx < pfcands->size(); ++trk2_idx) {
+      edm::Ptr<pat::CompositeCandidate> trk2_ptr(pfcands, trk2_idx);
+      if(!trk2_selection_(*trk2_ptr)) continue;
+      if (trk1_ptr->charge()== trk2_ptr->charge()) continue;      
+      std::vector <float> trk1_mass;
+      std::vector <float> trk2_mass;
+      
+      for ( auto & masses: mhypothesis){
```

Comment:
This may also benefit the phi -> KK. We just need an extra part like
if (doPhiToKK) {
  set both track mass = K_MASS
  kstarMass = kstarBarMass
}
Then we don't need to worry about trk1isK_trk2isPi or valid_mhypothesis for phi case

Reply:
Hi @sarafiorendi ,
So If I get correctly the comment you say to add a seperate branch for B0 and B0 bar right? Just the massess or everything should be doubled ? (honestly I find no reason for that, just to clarify). Thanks.
Hi @ottolau ,
I honestly do not see how adding them as seperate branches will help the phill but ok

Reply:
Dear @sarafiorendi ,

Many thanks for the claryfication. I think I got it. Instead of using the candidate i will just save Mass_BarMass and ffit_Mass fit_BarMass. Ok I will do it. thanks again